### PR TITLE
Whodidntvote

### PR DESCRIPTION
**Pull request by pjfranzini**
Implementation of Pivotal tracker feature 58522612

Testing: Method hasnt_voted of checkpoint Model is tested, but no integration tests have been added (there don't seem to be analytics integration tests in general?)

### CODE REVIEW

File: app/models/checkpoint.rb [hunk 1]
```diff
@@ -17,6 +17,30 @@ def ratings_count(start_time, end_time, score="all")
     end
   end
 
+  def hasnt_voted(start_time, end_time)
+    hasnt_voted_list = []
+    scoped_ratings = get_scoped_ratings(start_time, end_time) if self.ratings.any?
+
+    if !self.ratings.any? || !scoped_ratings.any?
+      hasnt_voted_list << "all"
```
Comment (aomran) [inline app/models/checkpoint.rb:25]: Do you think this should be here? If you take this out then the method will return an empty array which "programatically" means "no one voted". An array with "all" might be confusing as an API.

When someone calls `checkpoint.hasnt_voted` they can check with `nil?` and then do whatever they want with the result. If we include "all" then the returned array has something... but it's saying that nothing was returned. Do you see what I mean?

Comment (aomran) [inline app/models/checkpoint.rb:25]: I think I miss-understood what this line of code does. The use of `!` confuses me.

File: app/models/classroom.rb [hunk 2]
```diff
@@ -6,4 +6,14 @@ class Classroom < ActiveRecord::Base
 
   belongs_to :teacher
   validates :name, presence: true
+
+
+  def students_list
```
Comment (aomran) [inline app/models/classroom.rb]: If you're trying to get an array of ids then have a look at [Pluck](http://api.rubyonrails.org/classes/ActiveRecord/Calculations.html#method-i-pluck)

File: app/models/checkpoint.rb [hunk 3]
```diff
@@ -17,6 +17,30 @@ def ratings_count(start_time, end_time, score="all")
     end
   end
 
+  def hasnt_voted(start_time, end_time)
+    hasnt_voted_list = []
+    scoped_ratings = get_scoped_ratings(start_time, end_time) if self.ratings.any?
+
+    if !self.ratings.any? || !scoped_ratings.any?
```
Comment (aomran) [inline app/models/checkpoint.rb]: try `if self.ratings.empty? || scoped_ratings.empty?`. Does that read better to you (it does to me).

File: app/models/checkpoint.rb (hunk 1 above)
Comment (pjfranzini) [inline app/models/checkpoint.rb:25]: Maybe the method isn't well-named :s 
I didn't really think of it in terms of how "someone" would use it, just in terms of how I was using it -- which was to create labels of "Hasn't voted: all" if no-one had voted; "Hasn't voted: -- list of names" if some people haven't voted; nothing if everyone has voted.

Comment (aomran) [inline app/models/checkpoint.rb:25]: I understand it now.

File: app/models/checkpoint.rb (hunk 3 above)
Comment (pjfranzini) [inline app/models/checkpoint.rb]: ok

File: app/models/classroom.rb (hunk 2 above)
Comment (aomran) [inline app/models/classroom.rb]: So you can avoid this method entirely. And do `student_list = self.track.classroom.students.pluck(:id)`

Comment (pjfranzini) [inline app/models/classroom.rb]: Got it, ty

**Comment by aomran** [comment]
I suggest we move the `get_scoped_ratings` method into the `Rating` model. So in the `rating.rb` model we should create a class method like this:

``` ruby

def self.method_name(start_time, end_time)
  where({ created_at: start_time..end_time }).group(:student_id)
end
```

The `method_name` should be replaced with something that says two things:
- scoped by time
- only one vote per student

I can't think of a good name at the moment.

**Comment by aomran** [comment]
I'm going to fix the time and scope issues in a separate branch. This is good to go.

**Pull request merged**